R/plot.R

```diff
@@ -448,46 +453,92 @@ contour.stars = function(x, ...) {
 #' if (require(ggplot2)) {
 #'  ggplot() + geom_stars(data = r) + scale_fill_identity()
 #' }
-st_rgb = function(x, dimension = 3, use_alpha = dim(x)[dimension] == 4, maxColorValue = 255L, 
-		probs = c(0., 1.), stretch = FALSE) {
+#' r = st_rgb(x[,,,3:1],
+#' 		   probs = c(0.01, 0.99),
+#' 		   stretch = "percent")
+#' plot(r)
+#' r = st_rgb(x[,,,3:1],
+#' 		   probs = c(0.01, 0.99),
+#' 		   stretch = "histogram")
+#' plot(r)
+st_rgb <- function (x,
+					dimension = 3,
+					use_alpha = dim(x)[dimension] == 4,
+					maxColorValue = 255L,
+					probs = c(0, 1),
+					stretch = NULL) {
 	if (is.character(dimension))
 		dimension = match(dimension, names(dim(x)))
-	stopifnot(is.numeric(dimension), length(dimension)==1)
-	if (!dim(x)[dimension] %in% c(3,4))
-		stop(paste("number of bands along dimension", dimension, "should be 3 or 4"))
+	stopifnot(is.numeric(dimension), length(dimension) == 1)
+	if (!dim(x)[dimension] %in% c(3, 4))
+		stop(paste("number of bands along dimension", dimension,
+				   "should be 3 or 4"))
 	dims = setdiff(seq_along(dim(x)), dimension)
-	cutoff = function(x, probs) {
-		qs = if (all(probs == c(0., 1.)))
+	cutoff = function(x, probs, stretch.method = "percent") {
+		if(stretch.method == "percent"){
+			qs = if (all(probs == c(0, 1)))
 				range(x)
-			else
-				quantile(x, probs, na.rm = TRUE)
-		x = (x - qs[1])/(qs[2] - qs[1])
-		x[x > 1] = 1
-		x[x < 0] = 0
-		x * maxColorValue
+			else quantile(x, probs, na.rm = TRUE)
+			x = (x - qs[1])/(qs[2] - qs[1])
+			x[x > 1] = 1
+			x[x < 0] = 0
+			x * maxColorValue
+		} else if(stretch.method == "histogram"){
+			x = stats::ecdf(x)(x)
+			x * maxColorValue
+		} else {
+			x
+		}
 	}
-	if (stretch)
-		x = st_apply(x, dimension, cutoff, probs = probs)
+
+	if(is.null(stretch)) {
+		stretch.method = "none"
+		stretch = FALSE
+	}
+
+	if(is.logical(stretch)) {
+		if(stretch){
+			stretch.method = "percent"
+			message("Stretch as logical == TRUE, then percent clip by probs is applied")
+		} else {
+			maxColorValue = max(x[[1]])
```

Comment:
so `maxColorValue` when given by user is ignored here? Should this be documented?

Reply:
There is a bug in the current version of `st_rgb()`. The `maxColorValue` could return error when it is lower than `max(x[[]])`, this is because the `rgb()` function requires that values belong to range `c(0, maxColorValue)`. Some times data could come with other data types with larger max values. If it is intended to be a user "cutoff limit" then a new method must be included. To keep the input `maxColorValue`, I changed the line:

```r
maxColorValue = max(x[[]])
```
...to

```r
maxColorValue = max(maxColorValue, max(x[[1]]))
```
What do you think?

Reply:
If helps, it is the logic of the changes:

Start
- (NOT CHANGED) Preliminar chekings
- re-define function cutoff():
  - defined params: 
    - `x` a vector to stretch,
    - `probs` a vector of probabilities (min, max) of length == 2,
    - `stretch.method` one of c('percent', 'histogram')
  - Control: streching by method defined
    - if `percent`, stretch by percent clip to quantiles
    - if `histogram`, stretch bu histogram equalization
    - Any other character, no stretching but rescaling to `maxColorValue` (intended Never used)
  - return: a vector of rescaled values with/without stretching
- Controls: Preparing methods and stretching status
  - if `stretch = NULL`, convert to `stretch = FALSE`
  - if `stretch` is logical value, then 
    - if `stretch = TRUE` set `stretch.metdod = "percent"`,
    - else re-define `maxColorValue` to max(maxColorValue, max(x[[]])) 
  - if `stretch = <Any character not in c('percent', 'histogram')>`
    - set `stretch.method = 'percent'` and `stretch = TRUE`
  - else set `stretch.method = stretch` (value fromargument `stretch`), then set `stretch = TRUE`
- Control: Process stretching
  if `stretch == TRUE` cutoff by stretch.method and probs if apply
- (NOT CHANGED) Color creation and final processing
End

Reply:
Hi @edzer. So, can I mark as "solved" this conversation? Agree?

Reply:
Only after I merge - I hope to get back to this in a few days.

Reply:
LGTM!